### PR TITLE
✨ Gradle sourcefile locations

### PR DESCRIPTION
**Pull request by vgranados-dd**
## What does this PR do?
- New BuildGradleMatcher
  - Current state of the matcher is only able to match dependencies defined without version range nor variables
- Establish BuildGradleMatcher as the matcher for both GradleLockfileExtractor and GradleVerificationMetadataExtractor
- Update all unit tests related

## Additional Notes
- After checking with someone with more expertise on gradle
  - Version ranges are not widely used
  - Lockfile are not mandatory for gradle, and are only available in the latest versions (same for version ranges)
  - Gradle allows for a lot of customization (constant definition files, etc.)

### CODE REVIEW

File: internal/utility/fileposition/column.go [hunk 1]
```diff
@@ -4,7 +4,7 @@ import (
 	"github.com/google/osv-scanner/internal/cachedregexp"
 )
 
-var wordRe = cachedregexp.MustCompile(`[^\s(\n|\r|\r\n)]+`)
+var wordRe = cachedregexp.MustCompile(`[^\s\r\n]+`)
```
Comment (marcwieserdev) [inline internal/utility/fileposition/column.go:7]: why do we remove this possibility ? How much does it work with CR, LF, CRLF files ?

File: pkg/lockfile/match-build-gradle.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+	// lockfile (default, groovy)
+	sourcefile, err := lockfile.Open(fileName)
+	if err != nil {
+		// kotlin
+		sourcefile, err = lockfile.Open(fileName + ".kts")
+	}
+
+	// gradle verification metadata (<rootdir>/gradle/verification-metadata.xml)
+	relativePath := "../" + fileName
```
Comment (marcwieserdev) [inline pkg/lockfile/match-build-gradle.go:24]: Is this a gradle constraint ?

File: pkg/lockfile/match-package-json_test.go [hunk 3]
```diff
@@ -11,7 +11,7 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-var matcher = lockfile.PackageJSONMatcher{}
+var packageJSONMatcher = lockfile.PackageJSONMatcher{}
```
Comment (marcwieserdev) [inline pkg/lockfile/match-package-json_test.go:14]: Is this rename necessary ?

File: pkg/lockfile/parse-gradle-lock_test.go
Comment (marcwieserdev) [inline pkg/lockfile/parse-gradle-lock_test.go:1]: General comment on this, I don't see any tests with a location assertion, is it expected ?

File: pkg/lockfile/parse-gradle-verification-metadata_test.go
Comment (marcwieserdev) [inline pkg/lockfile/parse-gradle-verification-metadata_test.go:1]: Same comment than gradle-lock_test file

**Review by marcwieserdev — COMMENTED** [review]
Few questions about it, but otherwise it looks good, thanks !

**Comment by marcwieserdev** [comment]
Another comment just came in my mind, how does the matcher plays with variables ? It is pretty common to have `var xxx = "the-version-here"` and then `implementation "artifactId:${xxx}"`

File: pkg/lockfile/match-build-gradle.go (hunk 2 above)
Comment (vgranados-dd) [reply pkg/lockfile/match-build-gradle.go:24]: When the `verification-metadata.xml` is generated, is always inside a folder named `gradle`. This means that the `build.gradle` file (source file that we are looking for) relative path will be "../build.gradle":

- <rootdir>/
   - gradle/
      - verification-metadata.xml
   - build.gradle
   - gradle.lockfile

File: pkg/lockfile/match-package-json_test.go (hunk 3 above)
Comment (vgranados-dd) [reply pkg/lockfile/match-package-json_test.go:14]: As there will be more matchers (this PR is adding `buildGradleMatcher`) in the `lockfile_test` module, they will collide by name if all are named `matcher`

File: pkg/lockfile/parse-gradle-lock_test.go
Comment (vgranados-dd) [reply pkg/lockfile/parse-gradle-lock_test.go:1]: For those extractors that are using a matcher, during the extract process (while parsing the lock file) there is no location extraction, because it will be set during the matching process. That's why in these test files the method `expectPackagesWithoutLocations` is used, and the matcher is mocked (`matchers_helper.go`)
There is a specific test for the matcher (`match-package-json_test.go` and `match-build-gradle_test.go`) and would be great to have some kind of "integration test" that test both the extraction and the matching.

**Review by marcwieserdev — APPROVED** [review]
